### DISCUSSION TITLE
discussion: how should water and lava interact (1.16)

### DISCUSSION THREAD

**Discussion by Felicis** [Felicis/Water-Physics-Overhaul#20] [Ideas]
Hi y'all, thanks for joining the discussion :)

If I am not mistaked the (vanilla) minecraft fluid mixing behaviour is:
1. Lava source + water (flowing or source) => lava source is replaced with obsidian
2. flowing lava + flowing water => flowing lava is replaced with cobblestone 
3. flowing lava + water source => water source is replaced with stone

For WPO we have a different situation because 
- infinite water sources are impossible
- fluids conserve their mass (flowing fluid is not infinite)
- fluids move about in packets (a flowing fluid can have 1-7 fluid packets)
- fluids flow and equalize (can move far from starting position)

When naively applying the vanilla fluid mixing rules, we get for example the following bad situations:
1. one water packet could turn an entire lava lake into obsidian (see #12)
2. cobblestone generators consume 1-7 packets of lava per cobble (because the flowing lava is replaced)
3. stone generators consume 8 water packets per stone

I think these situations would be very weird, so I am asking myself how the cases 1, 2, and 3 could be fixed to make sense in vanilla minecraft. What are your thoughts and ideas?

**Comment by Felicis** [comment]
Idea for 1:
creating obsidian consumes the lava source (8 packets) and one water packet (125 mb). the water cost is negligible considering how little obsidian is usually required and how abundant water is, however it means one bucket of water can only create 8 obsidian.

**Reply by Felicis** [reply]
creating obsidian could also consume 8 water packets (a full bucket). that seems more balanced, considering how little obsidian one needs. 
This would also make creating obsidian harder, and creating cobble more easy.

**Reply by Felicis** [reply]
also together with my idea 2 (cobble = 1 water + 1 lava packet & displace lava) this would make it possible to create volcanoes by inserting some water into a large body of lava chamber with a chute on top, which would then explode in volume to create 8 cobble per 1 lava and 1 water block (4x volume increase)... the lava would start rising and finally flowing out the chute :volcano: :eyes:

**Reply by SnoutBug** [reply]
i think this system will never be intuitive, imagine breaking a block at a lava lake which then prevents you from ever creating obsidian because the fill height is less than 8 blocks now.
there should be a way to determine wether the liquid is flowing (or in a re-leveling state) or still, which can then be used to implement the vanilla behaviour.
the fill level should not be a factor in my eyes, since we don't have snow layer like behaviour for cobble (except for cobble slabs which would be an option for less than or equal to half filled water+lava combining into cobble)

**Reply by Felicis** [reply]
You're right :thinking:
although you could still create obsidian by pouring a full bucket of lava into water. Would that be very inconvenient?

**Reply by djmrFunnyMan** [reply]
> imagine breaking a block at a lava lake which then prevents you from ever creating obsidian because the fill height is less than 8 blocks now.

Not really. Once the top layer turns to cobble you can just mine the cobble and the lava below it will be instantly converted into obsidian (assuming the area is flooded with water).
> there should be a way to determine wether the liquid is flowing (or in a re-leveling state) or still, which can then be used to implement the vanilla behaviour.

I think this might be too complicated and resource intensive for a minor improvement vs other ways of solving this bug. And it honestly feels even less intuitive than doing it _my way_

**Reply by SnoutBug** [reply]
* > you can just mine the cobble
  * in my mind, that is not a solution to the issue. here the player will be stuck in a frustrating loop of only creating cobble if only clunky enough. also I noticed that it is possible to have a slanted liquid surface, especially for larger lakes. pouring a bucket of water will then yield some obsidian and some cobble. all this from a lake which appears to be level.
* >and the lava below it 
  * works only for the case where there is a lava lake available that is deeper than 1 block

> feels even less intuitive

in what way is the vanilla behaviour not intuitive? I just don't see how guessing whether the block is 8 layers filled or just 7 and below is more intuitive.

> too complicated and resource intensive

I don't know the code base but this at least sounds like the thing I am describing -> [code](https://github.com/Felicis/Water-Physics-Overhaul/blob/2e0dd79de128b94cffbecfb4831b8f7392bbf004/src/main/java/net/skds/wpo/mixins/block/AbstractBlockStateMixin.java#L41)

**Reply by djmrFunnyMan** [reply]
> in my mind, that is not a solution to the issue. here the player will be stuck in a frustrating loop of only creating cobble

I mean if you really want a lot of obsidian then just go to the end or flood the lava lake with a natural water lake instead of using water buckets. And just no what I said is still valid this situation won't be frustrating at all. Besides you can also just wall a section off or fill the the missing block to raise the lava level again or just don't mess it up in the first place. (side note, it's also not like mining a single block will decrease the level in the entire lava lake). 
And I don't understand how the rest of what you said is relevant.
> works only for the case where there is a lava lake available that is deeper than 1 block

nearly 100% are blocks deep though
> in what way is the vanilla behaviour not intuitive? I just don't see how guessing whether the block is 8 layers filled or just 7 and below is more intuitive.

Well it's not intuitive because it doesn't exist in this mod. In vanilla water can either be a source block or a flowing block but flowing blocks just don't exist here. I assume by flowing you'd mean arbitrarily determining if the block is moving based on the context but that's just confusing. Releveling and moving (flowing) is just an action the fluid block does it's not a constant state here.

And guesseing whether the block is 8 layers filled or just 7 is really not hard + you can use the F3 menu.

And in fact you just I just realized that you answered this yourself. It's more intiutive because you don't have to guess the level of a fluid block. It's an objective reality. Meanwhile determining whether the block is moving or whatever is a lot more confusing, has a lot of possible edge cases that would need to be fixed and harder to code.

Also in your scenario a tiny level 1 drop of lava would turn into an obisidan block if it wasn't moving.

**Comment by Felicis** [comment]
Idea for 2:
instead of replacing all 1-7 lava packets with cobble, the cobble gen could:
- consume 1 lava packet and 1 water packet per cobble
- create the cobble in the spot where the lava was
- push the remaining lava packets back behind the generated cobble block

**Reply by Felicis** [reply]
the question is what to do when the lava cannot be escape (build chimney, pour lava, pour water => if lava can not flow through water it can not escape). Should the lava be destroyed? should the cobble be created in place of the water and displace the water instead? Or depending which fluid can move?

**Comment by Felicis** [comment]
Idea for 3:
I have no easy idea how to distinguish stone and cobble generators. The issue is that I think that it is not easy to keep the water from becoming a source block (8 packets) without making cobble gens really hard to build. On the other hand the lava for a working cobble gen also has to be kept below 8 packets, or it makes obsidian, which is hard, too.... 
The first though that came to my mind is requiring a specific block below the lava or water, like a basalt gen. Something like packed ice, although that is a bit hard to come by in vanilla :thinking:

**Comment by djmrFunnyMan** [comment]
I think you might just be overthinking this. Here's is my occam's razor if you will:
(by a "full" block I mean one with 8 packets)

1. Full lava + Non-Full/Full water = Lava block removed and replaced with obsidian. Water block removed
2. Non-Full lava + Non-Full water = Lava block removed and replaced with cobble. Water block removed
3. Non-Full lava + Full water = Lava block removed. Water block removed and replaced with stone

In all cases it doesn't matter how many packets the block had it just gets removed/replaced on contact.

**Reply by djmrFunnyMan** [reply]
Oh also in case of contact with multiple blocks at a time:

1. All lava blocks in contact with the water get replaced with obisidian
2. All lava blocks in contact with the water get replaced with cobble
3. All water blocks in contact with the lava get replaced with stone

Example: (the water would get removed but this is just a visualisation)
![2023-06-19_13 09 51](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/5079ee94-ff1a-4752-8aba-27395382f197)

**Reply by Felicis** [reply]
thanks for the ideas dmjr :+1: 
For a start I have implemented your occams razor approach, but I've not been able test it yet (yesterday I've just managed to get the game to run for the first time after kicking out sdks-core. If you're interested I can start posting _very_ nightly versions somewhere here so you can play with them)

**Reply by djmrFunnyMan** [reply]
that'd be great!

**Reply by Dvarkian** [reply]
This approach seems the most intuitive to me, out of all the ones that have been suggested here so far.

P.S. Awesome work updating this mod Felicis. Thrilled to see that someone's taken it on!

**Comment by offbem** [comment]
Wouldn't the right thing with water in contact with lava be basalt?
